domains/client/relayer/src/worker.rs

```diff
@@ -57,74 +63,113 @@ pub async fn relay_consensus_chain_messages<Client, Block, SO>(
+                    domain_id,
+                    domain_block_number,
+                    domain_hash,
+                )?;
+            }
+
+            res
         },
         sync_oracle,
-        |chain_id, block_number| -> Result<bool, ApiError> {
-            let ChainId::Domain(domain_id) = chain_id else {
-                return Err(ApiError::Application(Box::from(
-                    "Should always be running under a Domain".to_string(),
-                )));
+        |consensus_block_number|
+         -> Result<Option<DomainExtraData<Block>>, ApiError> {
+            // since a parent mmr leaf is included in its child,
+            // we process the finalized block's parent instead since we know parent is implicitly finalized
+            // so we ensure the state of the parent is available here
+            let consensus_block_number = match consensus_block_number.checked_sub(&One::one()) {
+                None => return Ok(None),
+                Some(number) => number
             };
 
-            // short circuit if the domain state is unavailable to relay messages.
-            if !is_state_available(&domain_state_pruning, &domain_client, block_number) {
-                return Ok(false);
+            if !is_state_available(
+                &consensus_state_pruning,
+                &consensus_chain_client,
+                consensus_block_number,
+            ) {
+                return Ok(None);
             }
 
+            let consensus_hash_to_process = consensus_chain_client
+                .hash(consensus_block_number)?
+                .ok_or(ApiError::UnknownBlock(format!("Missing Hash for block number: {consensus_block_number:?}")))?;
             let api = consensus_chain_client.runtime_api();
-            let at = consensus_chain_client.info().best_hash;
-            Ok(api
-                .oldest_unconfirmed_receipt_number(at, domain_id)?
-                // ensure block number is at least the oldest tracked number
-                .map(|oldest_tracked_number| block_number >= oldest_tracked_number)
-                .unwrap_or(false))
+            let confirmed_domain_block =
+                api.latest_confirmed_domain_block(consensus_hash_to_process, domain_id)?;
+
+            if let Some((domain_block_number, domain_block_hash)) = confirmed_domain_block {
+                // short circuit if the domain state is unavailable to relay messages.
+                if !is_state_available(&domain_state_pruning, &domain_client, domain_block_number) {
+                    return Ok(None);
+                }
+
+                // check if this domain block is already relayed
+                if Relayer::fetch_domains_blocks_relayed_at(
+                    &domain_client,
+                    domain_id,
+                    domain_block_number,
+                )
+                    .contains(&domain_block_hash)
+                {
+                    return Ok(None);
+                }
```

Comment:
An edge case here, if the domain stops progressing (like not activity) then we only try to relay the XDM in the last confirmed domain block for once then mark the block processed and skip it afterward without retrying.

Reply:
True. We need a fallback mechanism that is not built in yet as I have mentioned in the other comment. Will handle that fallback mechanism when we do a cleanup of the aux storage